### PR TITLE
PHP 8.3 Support: Dynamic class constant fetch #6701

### PR DESCRIPTION
**Pull request by junichi11**
- https://github.com/apache/netbeans/issues/6701
- https://wiki.php.net/rfc/dynamic_class_constant_fetch

### PHP 8.3 Support: Add PHP 8.3 to the PhpVersion 

- https://wiki.php.net/rfc#php_83
- https://wiki.php.net/todo/php83

### PHP 8.3 Support: Dynamic class constant fetch (Part 1)

- Fix the grammar file (parser)
- Add the `PHP83UnhandledError`
- Add/Fix unit tests for the parser

Example:
```php
class Example {
    const CONSTANT = 'constant';
}
$constant = 'CONSTANT';
echo Example::{$constant};
```

PHP 8.3:

nb-php83-dynamic-class-constant-fetch-syntax

PHP 8.2:

nb-php83-dynamic-class-constant-fetch-php82

### PHP 8.3 Support: Dynamic class constant fetch (Part 2)

- Fix the `SemanticAnalysis`
- Fix the mark occurrences and the go to declaration features
- Add unit tests

### PHP 8.3 Support: Dynamic class constant fetch (Part 3)

- Fix the formatter
  - Add the option of the around spaces for the scope resolution operator (`::`)
  - Don't add spaces within braces (`{}`) e.g. {$example}
- Add unit tests

### PHP 8.3 Support: Dynamic class constant fetch (Part 4)

- Fix the code completion feature
- Add unit tests for the code completion and brace matching features

### PHP 8.3 Support: Dynamic class constant fetch (Part 5)

- Fix the `IntroduceSuggestion` and `PHP55UnhandledError`
- Add unit tests

### CODE REVIEW

File: php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java [hunk 1]
```diff
@@ -54,10 +62,24 @@ public Identifier getConstantName() {
         while (expression instanceof ExpressionArrayAccess) {
             expression = ((ExpressionArrayAccess) expression).getExpression();
         }
+        // can use dynamic constant names since PHP 8.3
+        // e.g. Foo::{$bar};
+        assert expression != null;
+        if (isDynamicName) {
+            // XXX create an identifier to avoid NPE, can improve this?
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java:69]: `XXX` - do we want to keep ii in sources?

File: php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java [hunk 2]
```diff
@@ -54,10 +62,24 @@ public Identifier getConstantName() {
...
+            if (expression instanceof ReflectionVariable) {
+                expression = ((ReflectionVariable) expression).getName();
+            }
+            expression = new Identifier(constant.getStartOffset(), constant.getEndOffset(), "{" + expression.toString() + "}"); // NOI18N
+        }
         assert expression instanceof Identifier;
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java:75]: Perhaps a bit better?

`assert expression instanceof Identifier : "Expected Identifier but got: " + expression.getClass().getName();`

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/PHPBracesMatcherTest.java [hunk 3]
```diff
@@ -522,7 +522,7 @@ public void testAttributeSyntax_05b() throws Exception {
                 + "    #[Class1(5)]\n"
                 + "    public function method~(#[Class1(6)] $param1, #[Class1('foo', 'bar', 7)] int $pram2@)^ {}\n"
                 + "\n"
-                + "@}^"
+                + "}"
```
Comment (tmysik) [inline php/php.editor/test/unit/src/org/netbeans/modules/php/editor/PHPBracesMatcherTest.java:525]: Why these changes exactly?

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/PHP55UnhandledError.java [hunk 4]
```diff
@@ -124,11 +124,15 @@ public void visit(StaticConstantAccess node) {
...
-                String constantName = constant.getName();
-                if ("class".equals(constantName.toLowerCase())) { //NOI18N
-                    createError(constant);
+            if (node.isDynamicName()) {
+                super.visit(node);
+            } else {
+                Identifier constant = node.getConstantName();
+                if (constant != null) {
+                    String constantName = constant.getName();
+                    if ("class".equals(constantName.toLowerCase())) { //NOI18N
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/PHP55UnhandledError.java:133]: Nitpick: I believe we should use some locale in `toLowerCase()` call.

File: php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java (hunk 1 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java:69]: Will remove `XXX`. Thanks!

File: php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java (hunk 2 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/parser/astnodes/StaticConstantAccess.java:75]: Yes, I'll change to it :) Thanks!

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/PHPBracesMatcherTest.java (hunk 3 above)
Comment (junichi11) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/PHPBracesMatcherTest.java:525]: Maybe, I forgot to remove them when I copied and pasted the previous test.
`@^~` are used to get positions. (e.g. `^` is caret position)
1. get positions using the `indexOf()` method
2. replace all `@^~` to empty string (`""`)

So, the extra markers did not affect unit tests.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/PHP55UnhandledError.java (hunk 4 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/PHP55UnhandledError.java:133]: Yes, I'll fix it. Thanks!

**Comment by junichi11** [comment]
@tmysik Fixed : https://github.com/apache/netbeans/compare/e53b9278c63011edf23dc592d66827e77d98edbe..7e705639b19ce57a0fe0984cdc8e91ac2f4272bc

Thank you for your review and your time :)

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/PHPBracesMatcherTest.java (hunk 3 above)
Comment (tmysik) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/PHPBracesMatcherTest.java:525]: I see, thanks for the explanation.

**Comment by tmysik** [comment]
@junichi11 Nothing to thank for, really. NetBeans users have to thank you, Junichi. Without you, there would not be any up-to-date PHP support in NetBeans at all! 👍
